.github/workflows/plugin_install.yml

```diff
@@ -3,120 +3,33 @@ name: Plugin Install
 on: [push, pull_request, workflow_dispatch]
 
 jobs:
-
-  linux-install:
-
-    runs-on: ubuntu-latest
-
+  plugin-install:
+    strategy:
+      matrix:
+        os: [windows-latest, ubuntu-latest]
+        jdk: [11, 17]
+    runs-on: ${{ matrix.os }}
     steps:
-      - uses: actions/checkout@v2
-
       - name: Set up JDK
         uses: actions/setup-java@v1
         with:
-          java-version: 11
+          java-version: ${{ matrix.jdk }}
 
-      - name: Build
-        run: ./gradlew clean assemble -Dbuild.snapshot=false
-
-      - name: Download OpenSearch Core
-        run: |
-          opensearch_version=`./gradlew properties -q | grep "opensearch_version:" | awk '{print $2}' | sed 's/-SNAPSHOT//g'`
-          wget https://ci.opensearch.org/ci/dbc/distribution-build-opensearch/$opensearch_version/latest/linux/x64/tar/builds/opensearch/dist/opensearch-min-$opensearch_version-linux-x64.tar.gz
-          tar -xzf opensearch-*.tar.gz
-          rm -f opensearch-*.tar.gz
-
-      - name: Move and rename security plugin for installation
-        run: mv build/distributions/opensearch-security-*.zip opensearch-security.zip
+      - name: Checkout Branch
+        uses: actions/checkout@v2
 
-      - name: Run OpenSearch with plugin
-        run: |
-          cat > os-ep.sh <<EOF
-          yes | opensearch-plugin install file:///docker-host/security-plugin.zip
-          chmod +x plugins/opensearch-security/tools/install_demo_configuration.sh
-          yes | plugins/opensearch-security/tools/install_demo_configuration.sh
-          chown 1001:1001 -R /opensearch
-          su -c "/opensearch/bin/opensearch" -s /bin/bash opensearch
-          EOF
-          docker build -t opensearch-test:latest -f- . <<EOF
-          FROM ubuntu:latest
-          COPY --chown=1001:1001 os-ep.sh /docker-host/
-          COPY --chown=1001:1001 opensearch-security.zip /docker-host/security-plugin.zip
-          COPY --chown=1001:1001 opensearch* /opensearch/
-          RUN chmod +x /docker-host/os-ep.sh
-          RUN useradd -u 1001 -s /sbin/nologin opensearch
-          ENV PATH="/opensearch/bin:${PATH}"
-          WORKDIR /opensearch/
-          ENTRYPOINT /docker-host/os-ep.sh
-          EOF
-          docker run --name ops -d -p 9200:9200 -p 9600:9600 -i opensearch-test:latest
+      - name: Assemble target plugin
+        run: ./gradlew assemble
 
-      - name: Sleep while OpenSearch finishes starting up
-        uses: whatnick/wait-action@v0.1.2
+      - name: Run Opensearch with A Single Plugin
+        uses: ./.github/actions/start-opensearch-with-one-plugin
         with:
-          time: '30s'
+          opensearch-version: 3.0.0
+          plugin-name: opensearch-security
+          plugin-start-script: install_demo_configuration
```

Comment:
Can these parameters be multi-line? I see an opportunity for user's of this action to run commands before node startup to alter the configuration as necessary. Perhaps we can have all of:

```
chmod +x plugins/${{ inputs.plugin-name }}/tools/install_demo_configuration.sh
yes | plugins/${{ inputs.plugin-name }}/tools/install_demo_configuration.sh
```

as this parameter and decouple this even more from the security plugin?

Reply:
I am not sure what you mean by be multi-line. Swapping those two lines into the workflow instead of the action is possible but I do not know that we want that. If we did this and ran it outside the current step in the action (splitting it in two) it would require another plugin to know the exact spot in the script where they wanted to inject their own version of this script call as well as know the specific path. If instead we just swap it into an input parameter, in the exact same spot it does not really save any overhead and makes the parameters messier. Likewise, you could extract this code snippet and make it a separate action but I do not know that doing so would save any other plugin any more work. They would still have take the existing action and modify it to match whatever paths and file names they use.

Reply:
I think there needs to be a way to append to the config file before starting up a node. The concept of the setup script is nice and general, but I think it's too tightly coupled with the security plugin which has a tools folder with scripts. I'm not sure if any other plugins have tools folders with scripts. 

What do you think about having a step before the action in the workflow to create a setup script? The example below creates a file called `setup.sh` which installs the demo configuration and also appends to `opensearch.yml` to allow rest API security config modification.

```
- name: Create Setup Script
run: |
        cat > setup.sh <<EOF
        chmod +x plugins/${{ inputs.plugin-name }}/tools/install_demo_configuration.sh
        yes | plugins/${{ inputs.plugin-name }}/tools/install_demo_configuration.sh
        echo plugins.security.unsupported.restapi.allow_securityconfig_modification: true >> config/opensearch.yml
        EOF
```

Then we can set `plugin-setup-script` as the new `setup.sh` file and if there needs to be actions with different configs in the future then all it would take is defining your own creation of the setup script.

Reply:
Thank you for your feedback and example. Sorry if I am being daft, but I am not sure I understand what you are requesting. I think you want to create a separate step that creates a new file `setup.sh` which holds the two lines for the linux install command and then another line that adds the security REST api to the opensearch.yml. If this is the case, I caution against this as we would have to have two parameters to action.yml because Windows cannot use the same commands and now we have removed the file name parameter and replaced it with a linux-specific script. 

If we do want to add functionality for modifying the opensearch.yml before running the action I would probably leave that to another issue as I am not sure what that would look like or how we would use it. I could be misunderstanding your intent however.

Reply:
Yes, some way of modifying the configuration before startup is needed. 

To try to clarify on why I think this is coupled with the security plugin is because `plugin-start-script` is used to run a script called `plugins/${{ inputs.plugin-name }}/tools/${{ inputs.plugin-start-script }}.sh` or `.bat` depending on the platform.

The security plugin has a `tools` directory with tools like the installation of the demo configuration, security admin and hash.

I just went through 3 other plugin repos:

- https://github.com/opensearch-project/observability
- https://github.com/opensearch-project/notifications
- https://github.com/opensearch-project/sql

and none have a `tools` directory in the same way.

Reply:
I separated the logic for handling a demo script path to only run if the there was demo path provided. Otherwise it will not run. That should mean that any other plugins can simply remove the script name from the action when they are setting things up.

Reply:
Sorry @cwperks but I still do not understand what you are hoping to see here. I know you have concerns about the re-usability of the action but I think that there is always going to be some variance between the different plugins. I do not believe that adding the options directly into the opensearch.yml is an option at the moment as we would have to do it between the time that opensearch installs the security plugin and the when it tries to run. This would mean that we would either have to split the action into two parts with a new step to run the demo script in the middle or just add two if-statemented steps for each OS inside the action itself. This seems more complicated then just writing a note that tells other plugins who do not use scripts to remove the lines. 

There is always going to be a bit of different between each code base unfortunately, as things get named in different ways etc. I think that this still offers a very strong starting point for other plugins to build or adapt to their own needs while also working as intended for the security repo. In the meantime, I can swap the line  `plugins/${{ inputs.plugin-name }}/tools/install_demo_configuration.sh` to be an entire path handed into the file instead of just a relative path containing the `tools` subdirectory. How does that sound? I am motivated to get this merged so please let me know what you would like done.

Reply:
I looked into this and do not think that it will be a trivial change to reformat things with the setup.sh file as you suggested. A number of problems arise because this workflow functions on both Linux using docker and Windows using a manual setup. The only straightforward solution I think may be possible is having two separate setup.sh files -- one for each OS. This complicates things however and may not be in-line with the spirit of the original workflow. 

I went ahead and looked at the security-dashboards installation action to see what you were thinking as I know you had previously referenced that as an example. For that, I do not think that the setup steps for used will be directly applicable to this workflow. While there are echos into opensearch.yml they look like they are because the dashboards repository already has a system in place for operating based off of the contents of opensearch.yml. It looks like the same script running used in this action is also used there. 

I am of the opinion that if we want to implement further changes to the way the action works we can try to come up with a uniform solution for the various actions as part of another issue. This code is still widely usable by other plugins with only minor modifications and for the time being, I believe it is easy enough for others to remove the  the logic for the the script running as they see fit.

Reply:
It can only be used by the security plug-in in the current form because of the structure of having scripts in the `tools` folder.

What I am thinking is adding steps in the workflow file (`plugin_install.yml`) to create a setup step for each platform.

`plugin_install.yml`
```
- name: Create Setup Script for Linux
   if: ${{ runner.os == 'Linux' }}
   run: |
        cat > setup.sh <<EOF
        chmod +x plugins/${{ inputs.plugin-name }}/tools/install_demo_configuration.sh
        yes | plugins/${{ inputs.plugin-name }}/tools/install_demo_configuration.sh
        echo plugins.security.unsupported.restapi.allow_securityconfig_modification: true >> config/opensearch.yml
        EOF

- name: Create Setup Script for Windows
   if: ${{ runner.os == 'Windows' }}
   shell: pwsh
   run: |
        echo. >> setup.bat
        echo 'y', 'y', 'N' | .\opensearch-${{ inputs.opensearch-version }}-SNAPSHOT\plugins\${{ inputs.plugin-name }}\tools\install_demo_configuration.bat >> setup.bat

- name: Run Opensearch with A Single Plugin
        uses: ./.github/actions/start-opensearch-with-one-plugin
        with:
          opensearch-version: 3.0.0
          plugin-name: opensearch-security
          plugin-setup-script: setup
          docker-host-plugin-zip: security-plugin.zip
```

and in the action:

`action.yml`
```
- name: Install Plugin into OpenSearch
   ...

- name: Setup OpenSearch for Linux
   run: `${{ inputs.plugin-setup-script}}`.sh

- name: Setup OpenSearch for Windows
   run: `${{ inputs.plugin-setup-script}}`.bat

- name: Run OpenSearch with plugin
   ....
```

Thank you for filing an issue for follow-up: https://github.com/opensearch-project/security/issues/2252

Reply:
I agree with @cwperks here (ty for the sample code). Making this modular and modifiable will give provide an opportunity for this workflow to be re-usable by other plugin and makes this flow much cleaner.